src/types.ts

```diff
@@ -1,6 +1,23 @@
 import * as revalidateEvents from './constants/revalidate-events'
 
-export type Fetcher<Data> = (...args: any) => Data | Promise<Data>
+export type Result<T = unknown> = T | Promise<T>
+
+export type Fetcher<
+  Data = unknown,
+  Args extends Key = Key
+> = Args extends (() => [infer R, ...infer K] | readonly [infer R, ...infer K])
```

Comment:
Why do we need `infer R, ...infer K` here instead of just one `...infer`?

Reply:
```typescript
useSWR(key: string | null)
useSWR(key: Record<any, any>)
useSWR(...keys: [any, any[]])
// typescript cant get the first and second inferred correctly by [...infer K]
// so i use a [any, unknown[]] here as a workaround here
// i cant give you a clear explanation for this because i don't know much about this part 😂
```